src/ClientData/include/ClientData/ProcessData.h

```diff
@@ -78,37 +53,43 @@ class ProcessData final {
   [[nodiscard]] bool is_64_bit() const;
   [[nodiscard]] const std::string& build_id() const;
 
-  void UpdateModuleInfos(absl::Span<const orbit_grpc_protos::ModuleInfo> module_infos);
+  void UpdateModuleInfos(absl::Span<const orbit_grpc_protos::ModuleInfo> module_infos,
+                         const ModuleIdentifierProvider& module_identifier_provider);
```

Comment:
Having to pass the `ModuleIdentifierProvider`s to these methods makes for a complicated interface. Moreover, the `ModuleInMemory`s contained in this class contain `ModuleIdentifier`s, so they are tied to a specific `ModuleIdentifierProvider`: it seems wrong to potentially allow to pass different `ModuleIdentifierProvider`s. I think this class should be rethought a bit. Maybe making it hold a pointer to the `ModuleIdentifierProvider` would be an improvement.

Reply:
>Maybe making it hold a pointer to the ModuleIdentifierProvider would be an improvement.

I fear that this is not going to work. `ProcessData` gets created in the session setup along with target configuration. It takes quite some calls and futures until it reaches orbit app. So at that point we don't have any idea about the module identifier provider.
On the other module data gets created in orbit app's constructor, so we need to know the identifier provider already at this point.

Reply:
That was an example, but your answer shows the following conceptual problem. If `ProcessData` is created before `ModuleIdentifierProvider` and knows nothing about it, then this connection between `ProcessData` and the `ModuleIdentifier`s in `ModuleInMemory` (from a specific `ModuleIdentifierProvider`) seems out of place. In other words, this class has two lives: first it represents the process as it comes from the session setup, and then afterwards it gets tied to a specific `ModuleIdentifierProvider`. So does `ProcessData` depends on a specific `ModuleIdentifierProvider` or not?

Maybe we accept that this is ok. But back to my original point: this ping-pong between `ProcessData` and `ModuleIdentifierProvider` is not ideal. `ModuleIdentifierProvider` is passed to compute `ModuleIdentifier`s inside the methods. But I think they should be computed from the outside. For example:
- `UpdateModuleInfos` and `AddOrUpdateModuleInfo` can be passed `ModuleInMemory`s directly, with the ids already set from the outside. They shouldn't know about `ModuleInfo`s.
- `FindModuleBuildIdsByPath` doesn't seem to be a responsibility of this class, as it doesn't need the address ranges of `ModuleInMemory`. Indeed, that's why the current implementation needs a set. Again, this class should deal with `ModuleInMemory`s and `ModuleIdentifier`s, not with parts of `ModuleInfo`s such as build ids/paths/filenames.
-  `FindModulesByFilename` can be passed `ModuleIdentifier`s directly, extracted from `ModuleIdentifierProvider` based on the filename. In general, this class should only use `ModuleIdentifier` and shouldn't know about filenames.

This should remove the dependencies of `ProcessData` from `ModuleInfo` and `ModuleIdentifierProvider`, which is desired because this class should primarily contain and not so much behavior that involves so many classes.

I'm sure I missed many details, but this is the general idea of what I originally meant with "this class should be rethought".

Reply:
So I managed to make process data own the provider. I guess this is better now. Follow up might be to get rid of some unrelated methods in process data (i'd argue that does not belong here).